### PR TITLE
Verify the storage root and tx trie root in runtime

### LINKED ISSUE #37: Runtime block validation should check transactions root
- [ ] Make block check transactions trie root
- [ ] Implement trie root (either external or in wasm)

### LINKED ISSUE #39: Runtime block validation should check state root
At present, the storage trie is not calculated during the execution of the runtime, so it's rather difficult to verify the storage root in the header. We'll have to introduce an external - `calculate_storage_root`  or whatever, which forces the storage root to be evaluated.

### PR DESCRIPTION
**Pull request by gavofyork**
This API only works with our in-memory stuff, but that's fine for PoC-1. We'll need to rethink this for PoC-2 when we have a more serious backend.

Closes #37 
Closes #39

### CODE REVIEW

File: executor/src/wasm_executor.rs [hunk 1]
```diff
@@ -136,90 +148,80 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
...
+		written as u32
+	},
+	ext_storage_root(result: *mut u8) => {
+		let r = this.ext.storage_root();
+		this.memory.set(result, &r[..]).map_err(|_| DummyUserError)?;
+	},
+	ext_enumerated_trie_root(values_data: *const u8, values_len: u32, lens_data: *const u32, lens_len: u32, result: *mut u8) => {
+		let values = (0..lens_len)
+			.map(|i| this.memory.read_primitive(lens_data + i * 4))
+			.collect::<::std::result::Result<Vec<u32>, DummyUserError>>()?
```
Comment (gavofyork) [inline executor/src/wasm_executor.rs]: anyone know of any better way of shortcircuiting a `Result` than `.collect()?.into_iter()`?

Comment (arkpar) [reply executor/src/wasm_executor.rs]: https://doc.rust-lang.org/1.0.0/std/result/fn.fold.html perhaps

File: executor/src/wasm_executor.rs [hunk 2]
```diff
@@ -136,90 +148,80 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
...
-					0
-				} else {
-					5
-				}
-			} else {
-				4
-			}
-		})()
+		let mut sig = [0u8; 64];
+		this.memory.get_into(sig_data, &mut sig[..]).map_err(|_| DummyUserError)?;
```
Comment (arkpar) [inline executor/src/wasm_executor.rs]: Previous version preserved error code here. Is it intentionally discarded now?

Comment (gavofyork) [reply executor/src/wasm_executor.rs]: yup - any attempt to read out of bounds memory is now a strict runtime error.

**Pull request merged**